### ISSUE TITLE
Compilation issue 

### ISSUE THREAD

**Issue by harishmurkal** [danielaparker/jsoncons#178]
Hi,

I am considering your software for the first time. I have taken the latest release Release 0.123.2.
I am getting this following error of re-definition. When fixing by removing one of the definition I'm getting other consequential errors. Just wanted to know if any flags as such available for overcoming this issue?

Thanks and Regards,
Harish M

Compilation log snippet:
==========
Scanning dependencies of target Nim_UT
[ 94%] Building CXX object CMakeFiles/Nim_UT.dir/imsgit_local/work/harism/reg/5g_core/ngnim/src/NIM_NfNotification.cxx.o
In file included from /imsgit_local/work/harism/reg/5g_core/ngnim/jsoncons/include/jsoncons/json_exception.hpp:13:0,
                 from /imsgit_local/work/harism/reg/5g_core/ngnim/jsoncons/include/jsoncons/basic_json.hpp:26,
                 from /imsgit_local/work/harism/reg/5g_core/ngnim/jsoncons/include/jsoncons/json.hpp:10,
                 from /imsgit_local/work/harism/reg/5g_core/ngnim/src/NIM_NfNotification.cxx:20:
/imsgit_local/work/harism/reg/5g_core/ngnim/jsoncons/include/jsoncons/unicode_traits.hpp:1482:12: error: specialization of ‘std::is_error_code_enum<jsoncons::unicons::conv_errc>’ after instantiation
     struct is_error_code_enum<jsoncons::unicons::conv_errc> : public true_type
            ^
**/imsgit_local/work/harism/reg/5g_core/ngnim/jsoncons/include/jsoncons/unicode_traits.hpp:1482:12: error: redefinition of ‘struct std::is_error_code_enum<jsoncons::unicons::conv_errc>’
In file included from /imsgit/Tools/tsp_tools/CONTAINER_TOOLSET/toolset_CENTOS/usr/include/c++/4.8.2/mutex:43:0,**
                 from /imsgit_local/work/harism/reg/ims_tools/grpc/protobuf/include/google/protobuf/stubs/mutex.h:33,
                 from /imsgit_local/work/harism/reg/ims_tools/grpc/protobuf/include/google/protobuf/stubs/common.h:52,
                 from /imsgit_local/work/harism/reg/5g_core/nimrpc/inc/NimDiscovery.pb.h:9,
                 from /imsgit_local/work/harism/reg/5g_core/nimrpc/inc/NimDiscovery.grpc.pb.h:22,
                 from /imsgit_local/work/harism/reg/5g_core/ngnim/inc/NIM_GrpcHandler.h:12,
                 from /imsgit_local/work/harism/reg/5g_core/ngnim/inc/NIM_Framework.h:16,
                 from /imsgit_local/work/harism/reg/5g_core/ngnim/src/NIM_NfNotification.cxx:6:
/imsgit/Tools/tsp_tools/CONTAINER_TOOLSET/toolset_CENTOS/usr/include/c++/4.8.2/system_error:54:12: error: previous definition of ‘struct std::is_error_code_enum<jsoncons::unicons::conv_errc>’
     struct is_error_code_enum : public false_type { };

**Comment by danielaparker** [comment]
When you say, "removing one of the definitions", do you mean removing these lines, in `unicode_traits.hpp`?

```c++
#ifndef JSONCONS_UNICONS_UNICODE_TRAITS_HPP
#define JSONCONS_UNICONS_UNICODE_TRAITS_HPP

#include <system_error>

// HERE
namespace std {
    template<>
    struct is_error_code_enum<jsoncons::unicons::conv_errc> : public true_type
    {
    };
    template<>
    struct is_error_code_enum<jsoncons::unicons::encoding_errc> : public true_type
    {
    };
}
#endif
```

That's the only place these symbols are defined, and I believe those are legal declarations. Similar declarations are used for error codes, e.g.

```c++
#ifndef JSONCONS_JSON_ERROR_HPP
#define JSONCONS_JSON_ERROR_HPP

#include <system_error>

namespace std {
    template<>
    struct is_error_code_enum<jsoncons::json_errc> : public true_type
    {
    };
}

#endif
```
Are you also getting redefinition errors for these?

I'll look into setting up a build project on CentOS using travis, and see if I can reproduce this.

The specializations of `is_error_code_enum` in the `std` namespace are a convenience, but the code could 
be rewritten without them, so if that's the only issue, we could make those changes and provide a 
compile time flag to suppress them. 

Daniel

**Comment by danielaparker** [comment]
I would need your help to pursue this further. Absent that, there's not much I can do.

**Comment by patternoia** [comment]
Dear Daniel,
I am getting very similar compilation error in c++ (Debian 4.9.2-10+deb8u2) 4.9.2. Compiling it in clang 8 (Apple LLVM version 8.0.0 (clang-800.0.42.1)) works just fine.
When I am commenting out the `is_error_code_enum<jsoncons::unicons::conv_errc> : public true_type` block, the compilation succeeds, I am not sure however if it will have any side effects.
I am using jsoncons version 0.122.0, upgrading to 0.133.0 does not help.

I am sending you the log files over the email.

**Comment by danielaparker** [comment]
Thanks, reopening this issue, will investigate. If it compiles, there shouldn't be any issue from commenting out the line, it's purpose is only to indicate that the type `jsoncons::unicons::conv_errc` is eligible for automatic conversion to `std::error_code`.

**Comment by danielaparker** [comment]
Hi patternoia,

Could you take master, and rebuild Debian 4.9.2? Let me know if you still have the compilation error?

Thanks,
Daniel

**Comment by patternoia** [comment]
Hi Daniel, this fixes it. Thanks a lot for quick response.

**Comment by danielaparker** [comment]
Resolved in v0.134.0
